src/QsCompiler/Transformations/ClassicallyControlled.cs

```diff
@@ -112,25 +89,27 @@ private TypeArgsResolution GetCombinedTypeResolution(TypeArgsResolution outer, T
                         && !TypedExpression.IsPartialApplication(expr.Item.Expression)
                         && call.Item1.Expression is ExpressionKind.Identifier)
                     {
-                        // We are dissolving the application of arguments here, so the call's type argument
-                        // resolutions have to be moved to the 'identifier' sub expression.
-
-                        var callTypeArguments = expr.Item.TypeArguments;
-                        var idTypeArguments = call.Item1.TypeArguments;
-                        var combinedTypeArguments = GetCombinedTypeResolution(callTypeArguments, idTypeArguments);
+                        var newCallIdentifier = call.Item1;
+                        var callTypeArguments = expr.Item.TypeParameterResolutions;
+                        var idTypeArguments = call.Item1.TypeParameterResolutions;
 
                         // This relies on anything having type parameters must be a global callable.
-                        var newCallIdentifier = call.Item1;
-                        if (combinedTypeArguments.Any()
-                            && newCallIdentifier.Expression is ExpressionKind.Identifier id
-                            && id.Item1 is Identifier.GlobalCallable global)
+                        if (
+                            newCallIdentifier.Expression is ExpressionKind.Identifier id
+                            && id.Item1 is Identifier.GlobalCallable global
+                            && (callTypeArguments.Any() || idTypeArguments.Any()))
                         {
+                            // We are dissolving the application of arguments here, so the call's type argument
+                            // resolutions have to be moved to the 'identifier' sub expression.
+                            var combined = CallGraph.TryCombineTypeResolutions(global.Item, out var combinedTypeArguments, idTypeArguments, callTypeArguments);
+                            QsCompilerError.Verify(combined, "failed to combine type parameter resolution");
```

Comment:
I think you had a ToDo for this, but this error should probably be generated by the combining logic, rather than here, in the calling context. That may remove the need to use the "Try-" method patter for the combining method allowing us to move the out-parameter to the proper return value.

Reply:
I think it seems fair to have as a QsCompilerError. If this fails, then that means that the syntax tree passed into this step was incorrect. Each rewrite step should be able to rely on that the given syntax tree is indeed valid, otherwise that's a huge burden and unnecessary overhead to implement full error handling in each step.

Reply:
I am not suggesting that we get rid of the check, but that it should be moved to the logic in the CallGraph, which has the logic that it is checking. The check doesn't verify any logic specific to the Classically Controlled.

Reply:
The method in the call graph specifically can also be used to validate part of the syntax tree. Hence, it won't raise a QsCompilerError but merely indicate if the validation fails (either via bool or via generated diagnstics). However, this invokation here indeed should never fail, hence raising a QsCompilerError here (serving the same purpose as Debug.Assert) seems appropriate. It is not a matter of QsCompilerError checking any logic of the Classically Controlled; it is a matter of detecting errors as early as we can and not siliently ignoring inconsistent state/input/etc. if there is an easy way to check it.